src/cpp/session/SessionSourceDatabaseSupervisor.cpp

```diff
@@ -116,8 +117,21 @@ FilePath sessionLockFilePath(const FilePath& sessionDir)
    return sessionDir.complete("lock_file");
 }
 
-// session dir lock (initialized by attachToSourceDatabase)
-FileLock s_sessionDirLock;
+// session dir lock (lock is acquired within 'attachToSourceDatabase()')
+boost::shared_ptr<FileLock> createSessionDirLock()
+{
+   bool isServer = session::options().programMode() == kSessionProgramModeServer;
+   FileLock::LockType lockType = isServer
+         ? FileLock::LOCKTYPE_LINKBASED
+         : FileLock::LOCKTYPE_ADVISORY;
+   return FileLock::create(lockType);
+}
+
+boost::shared_ptr<FileLock>& sessionDirLock()
+{
+   static boost::shared_ptr<FileLock> instance = createSessionDirLock();
+   return instance;
```

Comment:
@jjallaire I think you might have mentioned this in the RStudio PR -- any preference between returning a reference to the smart pointer vs. returning a reference to the actual lock?